### PR TITLE
Fixed #25968 -- Changed project/app templates to use a "py-tpl" suffix.

### PR DESCRIPTION
**Pull request by lamby**
Debian packages unconditionally byte-compile .py files on installation and
do not silence errors by design. Therefore, we need a way of shipping these
invalid .py files without a .py extension but ensuring that when we
template them, they end up as .py.

We don't special-case .py files so that the all the TemplateCommand
command-line options (eg. extra_files and extensions) still work entirely
as expected and it may even be useful for other formats too.

Signed-off-by: Chris Lamb <chris@chris-lamb.co.uk>

### CODE REVIEW

**Comment by timgraham** [comment]
[Florian suggests](https://groups.google.com/d/msg/django-developers/nPShxGatRuE/e4Wp0zC3BgAJ) that all Python template files use the suffix (even the valid ones). I'd suggest ".tpl" as the extension. A Trac ticket is also required.

**Comment by lamby** [comment]
Updated - moved to ``.tpl``. Will need to check that this still means the files are included (without ``__init__.py`` tools like ``find_packages`` etc may not "find" them)

**Comment by timgraham** [comment]
I think we need to document the behavior. This will be backwards incompatible for any custom templates that have files ending with ".tpl", right? Maybe there's some way to try to account for that, e.g. if a file has ".tpl" but not another extension?

**Comment by claudep** [comment]
@lamby is there a ticket yet for this patch?

**Comment by lamby** [comment]
@claudep There is now: https://code.djangoproject.com/ticket/25968

**Comment by claudep** [comment]
Most if not all test failures can be fixed by:
```
diff --git a/tests/admin_scripts/tests.py b/tests/admin_scripts/tests.py
index 755fb0b..a37a71e 100644
--- a/tests/admin_scripts/tests.py
+++ b/tests/admin_scripts/tests.py
@@ -181,7 +181,7 @@ class AdminScriptTestCase(unittest.TestCase):
                 pass
 
         conf_dir = os.path.dirname(upath(conf.__file__))
-        template_manage_py = os.path.join(conf_dir, 'project_template', 'manage.py')
+        template_manage_py = os.path.join(conf_dir, 'project_template', 'manage.py.tpl')
 
         test_manage_py = os.path.join(self.test_dir, 'manage.py')
         shutil.copyfile(template_manage_py, test_manage_py)
```

**Comment by lamby** [comment]
Updated, thanks :)

**Comment by timgraham** [comment]
The branch needs a rebase and (in case you missed it) my comment is still outstanding, "I think we need to document the behavior. This will be backwards incompatible for any custom templates that have files ending with ".tpl", right? Maybe there's some way to try to account for that, e.g. if a file has ".tpl" but not another extension?"

**Comment by lamby** [comment]
Rebased. Documentation to follow.

**Comment by lamby** [comment]
Rebased & documentation added.

**Comment by timgraham** [comment]
There's a remaining test failure: `project_template.test_settings.TestStartProjectSettings.test_middleware_classes_headers`

**Comment by timgraham** [comment]
I think Claude's suggestion to limit the potential for backwards-incompatibilities is a good one:
What about limiting chopping to ".py.tpl" => ".py"?

**Comment by lamby** [comment]
> What about limiting chopping to ".py.tpl" => ".py"?

Makes sense. Have rewritten the patch a little - by limiting it to ``.py`` it actually becomes more of a "search/replace" than simply a "stripping" one (or we require ugly special-casing of ``.py``). Docs updated to match.

(Test passing locally, will see what CI says...)

**Comment by timgraham** [comment]
The failing test only runs on Python 3: `project_template.test_settings.TestStartProjectSettings.test_middleware_classes_headers`

Please check docs and flake8 failures too.

**Comment by lamby** [comment]
> The failing test only runs on Python 3: project_template.test_settings.TestStartProjectSettings.test_middleware_classes_headers

Fixed.

> Please check docs and flake8 failures too.

Fixed.

File: docs/releases/1.9.2.txt [hunk 1]
```diff
@@ -21,3 +23,28 @@ Bugfixes
 * Fixed a regression that caused the incorrect day to be selected when opening
   the admin calendar widget for timezones from GMT+0100 to GMT+1200
   (:ticket:`24980`).
+* Fixed a regression in ``ConditionalGetMiddleware`` causing ``If-None-Match`` checks
+  to always return HTTP 200 (:ticket:`26024`).
```
Comment (claudep) [inline docs/releases/1.9.2.txt]: That's probably a merge artifact. Please simply remove those 2 lines (currently duplicated).

File: docs/releases/1.9.2.txt [hunk 2]
```diff
@@ -21,3 +23,28 @@ Bugfixes
...
+
+Due to changes to the project and application templates in 1.9, Python source
+code files with invalid syntax were shipped because of the use of Django
+templating syntax. This was not previously an issue because the Django-specific
+syntax was always within docstrings.
+
+This caused difficulties for packaging systems that unconditionally
+byte-compile ``*.py`` files, either failing or generating confusing warnings.
+
+To overcome this, a ``.tpl`` suffix has been added to these Python files  so
```
Comment (claudep) [inline docs/releases/1.9.2.txt]: double space between "files" and "so"

Comment (lamby) [inline docs/releases/1.9.2.txt]: Updated.

File: docs/releases/1.9.2.txt (hunk 1 above)
Comment (lamby) [inline docs/releases/1.9.2.txt]: Updated

**Comment by lamby** [comment]
Updated to match in-line comments.

File: docs/releases/1.9.2.txt [hunk 3]
```diff
@@ -21,3 +23,26 @@ Bugfixes
...
+
+This caused difficulties for packaging systems that unconditionally
+byte-compile ``*.py`` files, either failing or generating confusing warnings.
+
+To overcome this, a ``.tpl`` suffix has been added to these Python files so
+that they do not appear as ``.py`` files in the filesystem. The resulting
+``.py.tpl`` suffix is also replaced with ``.py`` upon rendering - for example,
+a template with the filename ``manage.py.tpl`` will be saved at ``manage.py``.
+
+If you are already using custom Python templates that end with ``.py.tpl``, in
```
Comment (timgraham) [inline docs/releases/1.9.2.txt]: This advice doesn't work as `filename.py.tpl.tpl` wouldn't be rewritten.

**Comment by timgraham** [comment]
Some edits: http://dpaste.com/3RKKDMH

**Comment by lamby** [comment]
@timgraham Thanks! All merged.

**Comment by timgraham** [comment]
Any response to the latest post on the mailing list thread about using py-tpl as a suffix?

**Comment by lamby** [comment]
> Any response to the latest post on the mailing list thread about using py-tpl as a suffix?

(just to be clear, the code now does that)

**Comment by timgraham** [comment]
I'm planning 1.9.2 for Feb. 1 so it would be great to get a consensus on that last question before then.

**Comment by claudep** [comment]
@lamby `.py.tpl` vs `.py-tpl`

**Comment by lamby** [comment]
@claudep @timgraham I like ``.py-tpl`` - it's immediately obvious that "something interesting" is going on, whilst it is possible for one to not notice ``.py.tpl``.

Shall I go ahead and update the patch?

**Comment by timgraham** [comment]
I'm in favor.

**Comment by lamby** [comment]
@timgraham @claudep Updated

**Comment by timgraham** [comment]
merged in abc0777b63057e2ff97eee2ff184356051e14c47, thanks!

**Pull request closed without merging**